SkyViewC3Service/Repositories/TankRepository.cs

```diff
@@ -0,0 +1,163 @@
+            return Task.Run<IEnumerable<LN2LevelCalibration>>(() => db.LN2LevelCalibrations.ToArray());
+        }
+        #endregion
+
+        #region TopTempCalibration
+        public async Task<TopTempCalibration> AddTopTempCalibrationAsync(TopTempCalibration calibration)
+        {
+            await AddCalibration(db.TopTempCalibrations, calibration);
+
+            return null;
+        }
+
+        public async Task<TopTempCalibration> RemoveTopTempCalibration(double reference)
+        {
+            await RemoveCalibration(db.TopTempCalibrations, reference);
+
+            return null;
+        }
+
+        public Task<IEnumerable<TopTempCalibration>> RetrieveAllTopTempCalibrations()
+        {
+            return Task.Run<IEnumerable<TopTempCalibration>>(() => db.TopTempCalibrations.ToArray());
+        }
+        #endregion
+
+        #region BottomTempCalibration
+        public async Task<BottomTempCalibration> AddBottomTempCalibrationAsync(BottomTempCalibration calibration)
+        {
+            await AddCalibration(db.BottomTempCalibrations, calibration);
+
+            return null;
+        }
+
+        public async Task<BottomTempCalibration> RemoveBottomTempCalibration(double reference)
+        {
+            await RemoveCalibration(db.BottomTempCalibrations, reference);
+
+            return null;
+        }
+
+        public Task<IEnumerable<BottomTempCalibration>> RetrieveAllBottomTempCalibrations()
+        {
+            return Task.Run<IEnumerable<BottomTempCalibration>>(() => db.BottomTempCalibrations.ToArray());
+        }
+        #endregion
+
+        #region ByPassTempCalibration
+        public async Task<ByPassTempCalibration> AddByPassTempCalibrationAsync(ByPassTempCalibration calibration)
+        {
+            await AddCalibration(db.ByPassTempCalibrations, calibration);
+
+            return null;
+        }
+
+        public async Task<ByPassTempCalibration> RemoveByPassTempCalibration(double reference)
+        {
+            await RemoveCalibration(db.ByPassTempCalibrations, reference);
+
+            return null;
+        }
+
+        public Task<IEnumerable<ByPassTempCalibration>> RetrieveAllByPassTempCalibrations()
+        {
+            return Task.Run<IEnumerable<ByPassTempCalibration>>(() => db.ByPassTempCalibrations.ToArray());
+        }
+        #endregion
+
+        #region Helper Functions
+        private async Task<T> AddCalibration<T>(DbSet<T> dbSet, T calibration) where T : Calibration
+        {
+            T[] calibrations = dbSet.Where(cal => cal.Reference == calibration.Reference).ToArray();
+            
+            if (calibrations.Length > 0)
+                calibrations[0].Value = calibration.Value;
+            else
+                await dbSet.AddAsync(calibration);
+            
+            await db.SaveChangesAsync();
+
+            return null;
+        }
+
+        private async Task<T> RemoveCalibration<T>(DbSet<T> dbSet, double reference) where T : Calibration
```

Comment:
hmm.. reference is unique ? we should check PK of Calibration Table. 
we should change primary key. 

Current Primary key  : (reference, value)
Later Primary key : reference